### ISSUE TITLE
SQLRelatedJoin with filter generates incorrect sql

### ISSUE THREAD

**Issue by yudytskiy** [sqlobject/sqlobject#187]
Hi! That's me again
Please correct me if I am wrong, but I get unexpected behavour.

Simpliest example:
We have Roles and Users and SQLRelatedJoin between them.

I would like to add property to the User to check if user is admin.

Code:

```
from sqlobject import (sqlhub, connectionForURI, SQLObject, SQLRelatedJoin, StringCol)

USER = 'some_db_user'
PASSWORD = 'some_db_password'
HOST = 'some_mysql_host_name_or_ip'
DB = 'some_db_name'
DRIVER = 'pymysql'
DATABASE_URI = f'mysql://{USER}:{PASSWORD}@{HOST}/{DB}?driver={DRIVER}'

class Role(SQLObject):
    name          = StringCol(length=100, alternateID=True)
    description   = StringCol(length=100, default=None)
    users         = SQLRelatedJoin('User')

class User(SQLObject):
    name          = StringCol()
    roles         = SQLRelatedJoin('Role')

    @property
    def is_admin(self):
        return self.roles.filter(Role.q.name == 'admin').count() != 0


if __name__ == '__main__':
    sqlhub.processConnection = connectionForURI(DATABASE_URI)
    Role.createTable(ifNotExists=True)
    User.createTable(ifNotExists=True)

    admin_role = Role(name='admin')
    user_role = Role(name='user')

    admin_user = User(name='This is Admin')
    admin_user.addRole(admin_role)
    admin_user.addRole(user_role)

    just_user = User(name='This is simple user')
    just_user.addRole(user_role)

    print(f'admin_user is {admin_user.is_admin} admin')
    print(f'just_user is {admin_user.is_admin} admin')

    select = just_user.roles.filter(Role.q.name == 'admin')
    print("sql = just_user.roles.filter(Role.q.name == 'admin') =", str(select))
    print('Result =', list(select))
    print('But we intended to select just admin roles...')

    print("Correct sql should be: SELECT _SO_SQLRelatedJoin_OtherTable.id, _SO_SQLRelatedJoin_OtherTable.name, _SO_SQLRelatedJoin_OtherTable.description FROM role  _SO_SQLRelatedJoin_OtherTable, role_user, user  _SO_SQLRelatedJoin_ThisTable WHERE (((_SO_SQLRelatedJoin_OtherTable.id = role_user.role_id) AND ((role_user.user_id = _SO_SQLRelatedJoin_ThisTable.id) AND (_SO_SQLRelatedJoin_ThisTable.id = 2))) AND ((_SO_SQLRelatedJoin_OtherTable.name) = ('admin')))")
```

The output:
admin_user is **True** admin
just_user is **True** admin
sql = just_user.roles.filter(Role.q.name == 'admin') = SELECT _SO_SQLRelatedJoin_OtherTable.id, _SO_SQLRelatedJoin_OtherTable.name, _SO_SQLRelatedJoin_OtherTable.description FROM role, role  _SO_SQLRelatedJoin_OtherTable, role_user, user  _SO_SQLRelatedJoin_ThisTable WHERE (((_SO_SQLRelatedJoin_OtherTable.id = role_user.role_id) AND ((role_user.user_id = _SO_SQLRelatedJoin_ThisTable.id) AND (_SO_SQLRelatedJoin_ThisTable.id = 2))) AND ((role.name) = ('admin')))
Result = [<Role 2 name='user' description=None>]
Correct sql should be: SELECT _SO_SQLRelatedJoin_OtherTable.id, _SO_SQLRelatedJoin_OtherTable.name, _SO_SQLRelatedJoin_OtherTable.description FROM role  _SO_SQLRelatedJoin_OtherTable, role_user, user  _SO_SQLRelatedJoin_ThisTable WHERE (((_SO_SQLRelatedJoin_OtherTable.id = role_user.role_id) AND ((role_user.user_id = _SO_SQLRelatedJoin_ThisTable.id) AND (_SO_SQLRelatedJoin_ThisTable.id = 2))) AND ((_SO_SQLRelatedJoin_OtherTable.name) = ('admin')))

**Comment by phdru** [comment]
I'm on vacation. Will be back in a week.

**Comment by phdru** [comment]
I'm here. Looking…

**Comment by phdru** [comment]
print(f'just_user is {admin_user.is_admin} admin')

This is your bug:

just_user is {**admin_user**.is_admin}

Before the fix for #185 the program (with fixed `just_user`) worked fine. So it seems I added a subtle bug in a4ee6cdb .

PS. `RelatedJoin` works fine and return `False` for the `just_user`.

**Comment by phdru** [comment]
Well, not enough SQL rewriting. The problem is in `.filter(Role.q.name` — the table `Role` is named explicitly, so it is added to the resulting SQL. It must be rewritten to `_SO_SQLRelatedJoin_OtherTable`. I have to think about that.

For now I advice to avoid explicit tables in `SQLRelatedJoin` filters and do filtering in Python:
```python
    def is_admin(self):
        return any((r.name == 'admin') for r in self.roles)
```

**Comment by phdru** [comment]
The rewrite is not always possible. Remember what a4ee6cd fixes? The case where `SQLRelatedJoin` joins a table with itself. In that case SQLObject couldn't guess to what alias to rewrite the filter and user must explicitly tells. So here is another fix for your problem, completely at the user level:
```python
    def is_admin(self):
        return self.roles.filter(RoleAlias.q.name == 'admin').count() != 0

RoleAlias = Alias(Role, '_SO_SQLRelatedJoin_OtherTable')
```

**Comment by phdru** [comment]
In a temporary commi c023d8f I added a test. The test passes before 3.10.2 but fails in 3.10.2. The other test passes/fails in opposite versions. Now I need to make both work at least in a simple case (two different tables).

**Comment by phdru** [comment]
Please check the commit ac28235 . It works for me. I added examples/tests. The last thing for me to think of is: how to raise an exception if a user filters aliased table without using an alias.

**Comment by yudytskiy** [comment]
Yes this commit fixes the issue.
I have mistake in code above.

print(f'just_user is {admin_user.is_admin} admin')
must be
print(f'just_user is {**just_user**.is_admin} admin')

So the output is:
admin_user is True admin
just_user is False admin
...

Both test code works with this commit.
Thanx!!!

**Comment by phdru** [comment]
Thanks for reporting!

**Comment by phdru** [comment]
Released 3.10.3 with the fix. I do not close the bug report yet — I want to think if it's possible to fix `relJoinCol.filter(Table.q.column)` by forcing the proper alias. Or at least to raise an error if `Alias` is not used.
